File: flutter/lib/busha_pay_result.dart
/// Result of a Busha Pay checkout.
///
/// Use pattern matching to handle each case:
/// ```dart
/// switch (result) {
///   case BushaPaySuccess(:final paymentId):
///     print('Payment $paymentId completed');
///   case BushaPayCancelled():
///     print('User cancelled');
///   case BushaPayError(:final message):
///     print('Error: $message');
/// }
/// ```
sealed class BushaPayResult {
  const BushaPayResult();
}

/// Payment completed successfully.
///
/// Full payment data (amounts, rates, timeline) is available when payment
/// was completed in the web checkout flow. When completed via the Busha app,
/// only [paymentId] and [status] are available.
///
/// Use [hasFullData] to check which fields are populated.
///
/// **Important:** Always verify the payment server-side via webhooks.
/// The client result is a UX hint, not the source of truth.
class BushaPaySuccess extends BushaPayResult {
  /// Payment request ID (e.g., 'PAYR_EhDmpnPwDSjO'). Always present.
  final String paymentId;

  /// Payment status (e.g., 'completed'). Always present.
  final String status;

  // Full data (web checkout flow only)

  /// Amount paid in source currency (e.g., '0.00017798').
  final String? sourceAmount;

  /// Source currency the user paid with (e.g., 'BTC').
  final String? sourceCurrency;

  /// Amount received in target currency (e.g., '13.42').
  final String? targetAmount;

  /// Target settlement currency (e.g., 'USDT').
  final String? targetCurrency;

  /// Originally requested amount.
  final String? requestedAmount;

  /// Currency of the requested amount.
  final String? currency;

  /// Exchange rate details.
  final Map<String, dynamic>? rate;

  /// Merchant information.
  final Map<String, dynamic>? merchantInfo;

  /// Payment timeline with step-by-step events.
  final Map<String, dynamic>? timeline;

  /// Full unmodified response from commerce-js.
  final Map<String, dynamic>? rawData;

  const BushaPaySuccess({
    required this.paymentId,
    required this.status,
    this.sourceAmount,
    this.sourceCurrency,
    this.targetAmount,
    this.targetCurrency,
    this.requestedAmount,
    this.currency,
    this.rate,
    this.merchantInfo,
    this.timeline,
    this.rawData,
  });

  /// Whether full payment data is available.
  ///
  /// `true` when payment completed in the web checkout (Path A).
  /// `false` when payment completed via the Busha app (Path B).
  bool get hasFullData => rawData != null;

  /// Creates a success result from a callback deep link (Path B — limited data).
  factory BushaPaySuccess.fromCallback({required String paymentId}) =>
      BushaPaySuccess(paymentId: paymentId, status: 'completed');

  /// Creates a success result from commerce-js onSuccess data (Path A — full data).
  factory BushaPaySuccess.fromCommerceJs(Map<String, dynamic> payload) {
    final data = payload['data'] as Map<String, dynamic>? ?? payload;

    return BushaPaySuccess(
      paymentId: data['id'] as String? ?? data['reference'] as String? ?? '',
      status: data['status'] as String? ?? 'completed',
      sourceAmount: data['source_amount'] as String?,
      sourceCurrency: data['source_currency'] as String?,
      targetAmount: data['target_amount'] as String?,
      targetCurrency: data['target_currency'] as String?,
      requestedAmount: data['requested_amount'] as String?,
      currency: data['currency'] as String?,
      rate: data['rate'] as Map<String, dynamic>?,
      merchantInfo: data['merchant_info'] as Map<String, dynamic>?,
      timeline: data['timeline'] as Map<String, dynamic>?,
      rawData: payload,
    );
  }

  @override
  String toString() => 'BushaPaySuccess(paymentId: $paymentId, status: $status, hasFullData: $hasFullData)';
}

/// Why a checkout ended without a completed payment.
enum BushaPayCancelledReason {
  /// The user dismissed the in-app chooser or web checkout sheet — close
  /// button, backdrop tap, or drag-to-dismiss.
  dismissed,

  /// The Busha app reported that the user explicitly rejected the payment.
  /// [BushaPayCancelled.paymentId] carries the payment request ID so you
  /// can reconcile server-side.
  rejected,

  /// The user returned to your app from the Busha app without a callback
  /// arriving. The payment outcome is **unverified** — it may still have
  /// succeeded. Always reconcile server-side (webhook / status API)
  /// before showing the user a final state.
  abandoned,
}

/// The checkout ended without a completed payment.
///
/// Inspect [reason] to tell *how* it ended. In particular,
/// [BushaPayCancelledReason.abandoned] does **not** mean the payment
/// failed — only that the SDK never received a result. Verify server-side.
class BushaPayCancelled extends BushaPayResult {
  /// How the checkout ended. Defaults to [BushaPayCancelledReason.dismissed].
  final BushaPayCancelledReason reason;

  /// The payment request ID. Present only when [reason] is
  /// [BushaPayCancelledReason.rejected] — the Busha app hands it back so
  /// the merchant can reconcile the rejected request.
  final String? paymentId;

  const BushaPayCancelled({this.reason = BushaPayCancelledReason.dismissed, this.paymentId});

  @override
  String toString() => 'BushaPayCancelled(reason: ${reason.name}, paymentId: $paymentId)';
}

/// An error occurred during checkout.
class BushaPayError extends BushaPayResult {
  /// Human-readable error message.
  final String message;

  /// Error code, if available.
  final String? code;

  const BushaPayError({required this.message, this.code});

  @override
  String toString() => 'BushaPayError(message: $message, code: $code)';
}

File: android/pay-android/src/main/kotlin/co/busha/pay/BushaPayResult.kt
package co.busha.pay

/**
 * Result of a Busha Pay checkout.
 *
 * Match exhaustively with `when`:
 * ```kotlin
 * when (result) {
 *     is BushaPayResult.Success -> println("Paid: ${result.paymentId}")
 *     is BushaPayResult.Cancelled -> println("Cancelled: ${result.reason}")
 *     is BushaPayResult.Error -> println("Error: ${result.message}")
 * }
 * ```
 */
sealed interface BushaPayResult {

    /**
     * Payment completed successfully.
     *
     * Full payment data (amounts, rates, timeline) is available when the
     * payment was completed in the web checkout flow. When completed via
     * the Busha app, only [paymentId] and [status] are populated — check
     * [hasFullData].
     *
     * **Important:** always verify the payment server-side via webhooks.
     * The client result is a UX hint, not the source of truth.
     */
    data class Success(
        /** Payment request ID (e.g. `"PAYR_EhDmpnPwDSjO"`). Always present. */
        val paymentId: String,
        /** Payment status (e.g. `"completed"`). Always present. */
        val status: String,
        val sourceAmount: String? = null,
        val sourceCurrency: String? = null,
        val targetAmount: String? = null,
        val targetCurrency: String? = null,
        val requestedAmount: String? = null,
        val currency: String? = null,
        val rate: Map<String, Any?>? = null,
        val merchantInfo: Map<String, Any?>? = null,
        val timeline: Map<String, Any?>? = null,
        /** Full unmodified payload from the web checkout. */
        val rawData: Map<String, Any?>? = null,
    ) : BushaPayResult {
        /**
         * `true` when full web-checkout data is present; `false` for
         * Busha-app callbacks, which carry only [paymentId] and [status].
         */
        val hasFullData: Boolean get() = rawData != null
    }

    /**
     * The checkout ended without a completed payment.
     *
     * Inspect [reason] to tell *how* it ended. In particular,
     * [BushaPayCancelledReason.ABANDONED] does **not** mean the payment
     * failed — only that the SDK never received a result. Verify
     * server-side before showing the user a final state.
     */
    data class Cancelled(
        /** How the checkout ended. */
        val reason: BushaPayCancelledReason = BushaPayCancelledReason.DISMISSED,
        /**
         * The payment request ID. Present only when [reason] is
         * [BushaPayCancelledReason.REJECTED] — the Busha app hands it
         * back so the merchant can reconcile the rejected request.
         */
        val paymentId: String? = null,
    ) : BushaPayResult

    /** An error occurred during checkout. */
    data class Error(
        /** Human-readable diagnostic message. */
        val message: String,
        /** Machine-readable error code, when available. */
        val code: String? = null,
    ) : BushaPayResult
}

/** Why a checkout ended without a completed payment. */
enum class BushaPayCancelledReason {
    /**
     * The user dismissed the in-app chooser or web checkout sheet —
     * close button, backdrop tap, or back press.
     */
    DISMISSED,

    /**
     * The Busha app reported that the user explicitly rejected the
     * payment. [BushaPayResult.Cancelled.paymentId] carries the request
     * ID so the merchant can reconcile server-side.
     */
    REJECTED,

    /**
     * The user returned to your app from the Busha app without a
     * callback arriving. The payment outcome is **unverified** — it may
     * still have succeeded. Always reconcile server-side.
     */
    ABANDONED,
}
